src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+        for res in resources.iter() {
+            match res.res_type {
+                IoType::Pio => {
+                    if self
+                        .pio_bus
+                        .insert(Range(res.addr.unwrap(), res.size), dev.clone())
+                        .is_some()
+                    {
+                        return Err(Error::Overlap);
+                    }
+                }
+                IoType::Mmio => {
+                    if self
+                        .mmio_bus
+                        .insert(Range(res.addr.unwrap(), res.size), dev.clone())
+                        .is_some()
+                    {
+                        return Err(Error::Overlap);
+                    }
+                }
+                IoType::PhysicalMmio => continue,
+            };
+        }
+        Ok(())
+    }
+
+    fn allocate_irq_resource(
+        &mut self,
+        interrupt: Option<IrqResource>,
+    ) -> Result<Option<IrqResource>> {
+        match interrupt {
+            Some(IrqResource(irq)) => {
+                // Allocate irq resource
+                let irq_num = self
+                    .resource
+                    .lock()
+                    .expect("failed to acquire lock")
+                    .allocate_irq(irq)
+                    .map_err(Error::IrqAllocate)?;
+                Ok(Some(IrqResource(Some(irq_num))))
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn free_irq_resource(&mut self, interrupt: Option<IrqResource>) {
+        match interrupt {
+            Some(IrqResource(irq)) => self
+                .resource
+                .lock()
+                .expect("failed to acquire lock")
+                .free_irq(irq),
+            None => return,
+        }
+    }
+
+    /// Register a new device with its parent bus and resources request set.
+    pub fn register_device(
+        &mut self,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: &mut Vec<IoResource>,
+        interrupt: Option<IrqResource>,
+    ) -> Result<()> {
+        // Reserve resources
+        let id = self.allocate_resources(resources)?;
+
+        // Register device resources
+        if let Err(Error::Overlap) = self.register_resources(dev.clone(), resources) {
+            return Err(Error::Overlap);
+        }
+
+        let name = dev.lock().expect("failed to acquire lock.").name();
+        let irq = self.allocate_irq_resource(interrupt)?;
+
+        // Set the allocated resource back
+        dev.lock()
+            .expect("Failed to acquire lock.")
+            .set_resources(resources, irq);
+
+        let descriptor = self.device_descriptor(id, name, dev, parent_bus, resources.to_vec(), irq);
+
+        // Insert bus/device to DeviceManager with parent bus
+        self.insert(descriptor)
+    }
+
+    /// Unregister a device from `DeviceManager`.
+    pub fn unregister_device(&mut self, instance_id: u32) -> Result<()> {
```

Comment:
How does the caller (The VMM) know which id to pass to `unregister_device` ? Shouldn't we return the device allocated id from `register_device`?

Reply:
@sameo, It seems this function would be called e.g. unhotplug. Otherwise, we won't do that right? If so, there needs a place to store and an API which might be supported later to list the device tree. So people would know id.

Reply:
@liujing2 I don't think we should assume any particular use case here. I should be able to register/unregister a device from the device manager at will, for example for testing. Since the handle for registering a device is its newly allocated id, we should return it from the register call and let the caller store id for calling unregister with it.

Reply:
Ok, got it. Will return the instance id for vmm managing in update.